### PR TITLE
Filter library by source

### PR DESCRIPTION
**Pull request by simakover**
Added option to filter library by sources

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt [hunk 1]
```diff
@@ -268,6 +268,8 @@ class PreferencesHelper(val context: Context) {
 
     fun filterTracking(name: Int) = flowPrefs.getInt("${Keys.filterTracked}_$name", ExtendedNavigationView.Item.TriStateGroup.State.IGNORE.value)
 
+    fun filterSources(name: Int) = flowPrefs.getInt("${Keys.filterSource}_$name", ExtendedNavigationView.Item.TriStateGroup.State.IGNORE.value)
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt:271]: Dont use Int for the name, source id's should always be Long

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt [hunk 2]
```diff
@@ -173,6 +173,9 @@ class LibraryPresenter(
         // SY -->
         val filterStarted = preferences.filterStarted().get()
         val filterLewd = preferences.filterLewd().get()
+        val inSources = sourceManager.getCatalogueSources().associate {
+            Pair(it.id.toInt(), preferences.filterSources(it.id.toInt()).get())
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt:177]: I generally dont use the Pair constructor, use the `to` keyword like this
`it.id to preferences.filterSources(it.id).get()`
Also make sure to use Long's

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt [hunk 3]
```diff
@@ -243,6 +246,21 @@ class LibraryPresenter(
...
         }
+
+        val filterFnSource: (LibraryItem) -> Boolean = source@{ item ->
+            val containsExclude = inSources.filterValues { it == State.EXCLUDE.value }
+            val containsInclude = inSources.filterValues { it == State.INCLUDE.value }
+
+            if (!containsExclude.any() && !containsInclude.any()) return@source true
+
+            val include = containsInclude.containsKey(item.manga.source.toInt())
+            val exclude = containsExclude.containsKey(item.manga.source.toInt())
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt:257]: Remove .toInt()

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibrarySettingsSheet.kt [hunk 4]
```diff
@@ -101,24 +103,37 @@ class LibrarySettingsSheet(
             // SY -->
             private val started = Item.TriStateGroup(R.string.started, this)
             private val lewd = Item.TriStateGroup(R.string.lewd, this)
+            private val sourceFilters: Map<Int, Item.TriStateGroup>
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibrarySettingsSheet.kt:106]: Use a Long

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibrarySettingsSheet.kt [hunk 5]
```diff
@@ -101,24 +103,37 @@ class LibrarySettingsSheet(
...
                         if (size > 1) list.add(Item.Header(R.string.action_filter_tracked))
                         list.addAll(trackFilters.values)
-                        items = list
                     }
+
+                // SY -->
+                sourceManager.getCatalogueSources().let { catalogue ->
+                    val size = catalogue.size
+                    sourceFilters = catalogue.associate {
+                        Pair(it.id.toInt(), Item.TriStateGroup(it.id.toInt(), this))
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibrarySettingsSheet.kt:130]: Refer to my other comment about pairs

File: app/src/main/java/eu/kanade/tachiyomi/widget/ExtendedNavigationView.kt [hunk 6]
```diff
@@ -269,7 +272,18 @@ open class ExtendedNavigationView @JvmOverloads constructor(
...
+
+                    // SY -->
+                    try {
+                        holder.text.setText(item.resTitle)
+                    } catch (e: Exception) {
+                        val sourceManager: SourceManager = Injekt.get()
+                        sourceManager.getCatalogueSources().filter { it.id.toInt() == item.resTitle }.map {
+                            holder.text.text = it.name
+                        }
+                    }
```
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/widget/ExtendedNavigationView.kt:284]: I wouldnt allow this, maybe make a different Holder for sources that returns the source name

**Review by jobobby04 — CHANGES_REQUESTED** [review]
I am not sure if I will even merge this, you can already search by source so there isnt really a benefit. Otherwise here are my general comments on the PR

**Comment by simakover** [comment]
@jobobby04 made int- > long changes, but dont understand about holder. it is have same functionality, why and how make new?

**Pull request closed without merging**